### PR TITLE
App accelerator #9: Valve Monitor

### PR DESCRIPTION
**Pull request by tjvantoll**
Monitor and control the open/closed state of a valve, and measure liquid flow state.

## Ticket(s)

https://trello.com/c/MON451h1/66-nf9-valve-monitor

### CODE REVIEW

File: valve-monitor/web-app/.devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// For format details, see https://aka.ms/devcontainer.json.
+{
+  "name": "Template Reference Web App Development Dockerfile",
```
Comment (paigen11) [inline valve-monitor/web-app/.devcontainer/devcontainer.json:3]: update this name

File: valve-monitor/web-app/.env.example [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+# If your company uses servers in a different region of the world.
+AZURE_REGION=eastus
+
+# Where will the site be viewable by users? (This will affect SSL certs.)
+# This is safe to leave as-is for deploying on Azure.
+SITE_DNS="$AZURE_SUBDOMAIN.$AZURE_REGION.azurecontainer.io"
+
+# Change if you deploy multiple versions of the site on one docker hub.
+SITE_DOCKER_IMAGE="$DOCKER_HUB_ACCOUNT/$APP_ID-site"
```
Comment (paigen11) [inline valve-monitor/web-app/.env.example:77]: remove this

File: valve-monitor/web-app/.prettierignore [hunk 3]
```diff
@@ -0,0 +1,5 @@
+/coverage
+/build
+/node_modules
+
+/package-lock.json
```
Comment (paigen11) [inline valve-monitor/web-app/.prettierignore:5]: add yarn-lock

File: valve-monitor/web-app/.vscode/settings.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    "statusBarItem.remoteBackground": "#4257fd",
+    "statusBarItem.remoteForeground": "#e7e7e7",
+    "tab.activeBorder": "#7484fe",
+    "titleBar.activeBackground": "#4257fd",
+    "titleBar.activeForeground": "#e7e7e7",
+    "titleBar.inactiveBackground": "#4257fd99",
+    "titleBar.inactiveForeground": "#e7e7e799"
+  },
+  "peacock.color": "#d753d6",
+  "peacock.remoteColor": "#4257fd"
```
Comment (paigen11) [inline valve-monitor/web-app/.vscode/settings.json:25]: remove these

File: valve-monitor/web-app/.vscode/tasks.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    {
+      "type": "npm",
+      "script": "dev",
+      "options": {
+        "env": {
+          "REACT_EDITOR": "code"
+        }
+      },
+      "problemMatcher": [],
+      "label": "Template reference app (local dev)",
```
Comment (paigen11) [inline valve-monitor/web-app/.vscode/tasks.json:13]: update label

File: valve-monitor/web-app/Dockerfile.prod [hunk 6]
```diff
@@ -0,0 +1,59 @@
+# You can build and run with:
+#   docker build . -t blues:reference-web-app &&
```
Comment (paigen11) [inline valve-monitor/web-app/Dockerfile.prod:2]: update this name?

File: valve-monitor/web-app/deploy.sh [hunk 7]
```diff
@@ -0,0 +1,145 @@
+#! /usr/bin/env bash
```
Comment (paigen11) [inline valve-monitor/web-app/deploy.sh:1]: remove azure from here

File: valve-monitor/web-app/docker-compose.azure.yml [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# docker-compose.azure.yml
```
Comment (paigen11) [inline valve-monitor/web-app/docker-compose.azure.yml:1]: remove

File: valve-monitor/web-app/environment.d.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+// Declare process.ENV types
+declare global {
+  namespace NodeJS {
+    interface ProcessEnv {
+      HUB_AUTH_TOKEN: string;
```
Comment (paigen11) [inline valve-monitor/web-app/environment.d.ts:5]: update these env vars

File: valve-monitor/web-app/src/components/presentation/deviceEventInfo.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
+/* eslint-disable import/prefer-default-export */
```
Comment (paigen11) [inline valve-monitor/web-app/src/components/presentation/deviceEventInfo.ts:1]: remove this file

File: valve-monitor/web-app/src/pages/api/device/[deviceUID]/valve-monitor.ts [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+import { StatusCodes } from "http-status-codes";
+import { ErrorWithCause } from "pony-cause";
+import { HTTP_STATUS } from "../../../../constants/http";
+import { services } from "../../../../services/ServiceLocatorServer";
+
+interface ValidRequest {
+  deviceUID: string;
+  valveMonitorConfig?: object;
+  name?: string;
+  state?: string;
```
Comment (paigen11) [inline valve-monitor/web-app/src/pages/api/device/[deviceUID]/valve-monitor.ts:12]: update to valveState to be more precise

File: valve-monitor/web-app/src/pages/api/log.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable no-console */
+export const serverLogError = console.error;
+export const serverLogInfo = console.info;
+
+// message 50% (500/1000)
+// message 40% (400/1000)
+// etc.
+export function serverLogProgress(
```
Comment (paigen11) [inline valve-monitor/web-app/src/pages/api/log.ts:8]: remove this function

File: valve-monitor/web-app/src/pages/index.tsx [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+    valveMonitorDevicesError &&
+    getErrorMessage(valveMonitorDevicesError.message);
+
+  const valveMonitorDevices = valveMonitorDeviceList;
+
+  const router = useRouter();
+  const queryClient = useQueryClient();
+
+  // refresh the page
+  const refreshData = async () => {
```
Comment (paigen11) [inline valve-monitor/web-app/src/pages/index.tsx:53]: is this function ever being used outside of `refreshDataAndInvalidateCache`?

File: valve-monitor/web-app/src/services/notehub/AxiosHttpNotehubAccessor.ts [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+    const endpoint = `${this.hubBaseURL}/v1/projects/${this.hubProjectUID}/fleets/${fleetUID}/environment_variables`;
+    try {
+      const resp = await axios.get(endpoint, { headers: this.commonHeaders });
+      return resp.data as NotehubEnvVarsResponse;
+    } catch (e) {
+      throw this.errorWithCode(e);
+    }
+  }
+
+  async setEnvironmentVariablesByFleet(
```
Comment (paigen11) [inline valve-monitor/web-app/src/services/notehub/AxiosHttpNotehubAccessor.ts:60]: add this to nextjs template

File: valve-monitor/web-app/src/services/notehub/AxiosHttpNotehubAccessor.ts [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    }
+    return errorCode;
+  }
+
+  errorWithCode(e: unknown): Error {
+    const errorCode = this.httpErrorToErrorCode(e);
+    return getError(errorCode, { cause: e as Error });
+  }
+
+  async setEnvironmentVariablesByDevice(
```
Comment (paigen11) [inline valve-monitor/web-app/src/services/notehub/AxiosHttpNotehubAccessor.ts:99]: add this to nextjs template

File: valve-monitor/web-app/src/services/notehub/NotehubAccessor.ts [hunk 16]
```diff
@@ -0,0 +1,27 @@
+import NotehubEnvVars from "./models/NotehubEnvVars";
+import NotehubFleets from "./models/NotehubFleets";
+import NotehubEnvVarsResponse from "./models/NotehubEnvVarsResponse";
+
+// An interface for accessing Notehub APIs
+interface NotehubAccessor {
+  getFleetsByDevice: (hubDeviceUID: string) => Promise<NotehubFleets>;
+  getDeviceEnvVars: (hubDeviceUID: string) => Promise<NotehubEnvVars>;
+  getEnvironmentVariablesByFleet: (
```
Comment (paigen11) [inline valve-monitor/web-app/src/services/notehub/NotehubAccessor.ts:9]: add this to nextjs template

File: valve-monitor/web-app/src/services/notehub/models/NotehubEnvVarsResponse.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
+import NotehubEnvVars from "./NotehubEnvVars";
```
Comment (paigen11) [inline valve-monitor/web-app/src/services/notehub/models/NotehubEnvVarsResponse.ts:1]: add this to nextjs template

File: valve-monitor/web-app/src/services/notehub/models/NotehubFleets.ts [hunk 18]
```diff
@@ -0,0 +1,11 @@
+interface NotehubFleets {
```
Comment (paigen11) [inline valve-monitor/web-app/src/services/notehub/models/NotehubFleets.ts:1]: add this to nextjs template

File: valve-monitor/web-app/src/services/prisma-datastore/schema.prisma [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+  provider = "postgresql"
+  url      = env("DATABASE_URL")
+}
+
+model Project {
+  id         Int       @id @default(autoincrement())
+  createdAt  DateTime  @default(now()) @map("created_at")
+  projectUID String    @map("project_uid") @unique
+  name       String    @db.VarChar(50)
+  fleets     Fleet[]
```
Comment (paigen11) [inline valve-monitor/web-app/src/services/prisma-datastore/schema.prisma:15]: add this to nextjs template

File: valve-monitor/web-app/src/styles/Card.module.scss [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+    font-size: 12px;
+  }
+
+  :global .ant-input-number {
+    border-color: $bw-color--grey3;
+  }
+  :global .ant-input-number:focus,
+  :global .ant-input-number:hover {
+    border-color: $bw-color--blue3;
+  }
```
Comment (paigen11) [inline valve-monitor/web-app/src/styles/Card.module.scss:53]: combine

File: valve-monitor/web-app/src/styles/Header.module.scss [hunk 21]
```diff
@@ -0,0 +1,49 @@
+@import "_variables.scss";
+
+.header {
+  display: flex;
+  justify-content: space-between;
+  padding: 0 32px !important;
+
+  :global .ant-layout-header {
```
Comment (paigen11) [inline valve-monitor/web-app/src/styles/Header.module.scss:8]: needed still?

File: valve-monitor/web-app/src/styles/ValveMonitorTable.module.scss [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+    margin: auto;
+  }
+
+  :global .ant-input-number {
+    width: 70px;
+    margin: auto;
+  }
+
+  :global .anticon-warning {
+    color: red;
```
Comment (paigen11) [inline valve-monitor/web-app/src/styles/ValveMonitorTable.module.scss:69]: use color variable

File: valve-monitor/web-app/src/styles/ValveMonitorTable.module.scss [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    :global .editable-button {
+      height: default;
+      white-space: nowrap;
+      overflow: hidden;
+    }
+
+    :global .ant-input-number {
+      margin: 0 0 0 16px;
```
Comment (paigen11) [inline valve-monitor/web-app/src/styles/ValveMonitorTable.module.scss:165]: combine

**Review by paigen11 — APPROVED** [review]

File: valve-monitor/web-app/src/pages/api/device/[deviceUID]/valve-monitor.ts [hunk 24]
```diff
@@ -57,16 +57,17 @@ function validateRequest(
...
 
 async function performPostRequest({
   deviceUID,
   valveMonitorConfig,
   name,
-  valveState,
+  state,
 }: ValidRequest) {
   const app = services().getAppService();
+  console.log("valve monitor file ", state);
```
Comment (tjvantoll) [inline valve-monitor/web-app/src/pages/api/device/[deviceUID]/valve-monitor.ts:70]: I think you accidentally left this in there.

Comment (paigen11) [reply valve-monitor/web-app/src/pages/api/device/[deviceUID]/valve-monitor.ts:70]: Thanks, removed it again

**Pull request merged**